src/Microsoft.AspNet.Diagnostics/StatusCodePagesMiddleware.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Http;
+
+namespace Microsoft.AspNet.Diagnostics
+{
+    public class StatusCodePagesMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly StatusCodePagesOptions _options;
+
+        public StatusCodePagesMiddleware(RequestDelegate next, StatusCodePagesOptions options)
+        {
+            _next = next;
+            _options = options;
+            if (_options.HandleAsync == null)
+            {
+                throw new ArgumentException("Missing HandleAsync implementation.");
+            }
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            await _next(context);
+
+            // Do nothing if a response body has already been provided.
+            if (context.Response.HeadersSent
+                || context.Response.StatusCode < 400
+                || context.Response.StatusCode >= 600
+                || context.Response.ContentLength.HasValue
+                || !string.IsNullOrEmpty(context.Response.ContentType))
```

Comment:
would be nice to have a single thing you can call to turn it off, so it can be slapped in `ResourceFilter` for api controllers.

Is the logic as simple as?

```C#
if (StatusCode >= 400)
{
   Respons.Body.Flush();
}
```

Because looks like any other way, actually starts sending stuff that is not necessary. Can we reconsider having an explicit turn-off feature?